### ISSUE TITLE
inputting custom floor plans

### ISSUE THREAD

**Issue by foojayyy** [MIT-SPARK/MiDiffusion#9]
Hello,

Thank you for sharing the excellent work on MiDiffusion! After reading the documentation, I have some questions regarding the use of custom input floor plans and would like to seek your guidance:

Does it support custom floor plans?
Does MiDiffusion support directly inputting custom floor plans for 3D scene generation? If so, what are the required data formats or specifications for the input?

Thank you for your time and assistance!

**Comment by SiyiHu** [comment]
We have tested MiDiffusion with custom floor plan using PointNet feature extractor, but it should also work with ResNet. 

The easiest way is to format the input data (which can contain just the floor plan and no objects) in the same way as preprocessed 3D-FRONT datasets. Then you can update `dataset_directory` in the config file to a `<custom_data_dir>` and run `scripts/generate_results.py <weight_file> --config_file config/<your_updated_config>.yaml`. Your custom data directory should look like this:
```
<ThreedFront_dir>/output/3d_front_processed/<custom_data_dir>/
├── dataset_stats.txt
├── room_1/
    ├── boxes.npz
    ├── room_mask.png
├── room_2/
    ├── boxes.npz
    ├── room_mask.png
```

You should copy `dataset_stats.txt` file from the preprocessed 3D-FRONT dataset directory (i.e. training data). This file contains upper and lower bounds for all geometric features (e.g. "bounds_translations"). These bounds are used to normalize raw input data and de-scale the predicted results. Hence, you need to pay attention to `"bounds_fpbpn"` ([x_min, y_min, dx_min, dy_min, x_max, y_max, dx_max, dx_max]), which is computed by `ThreedFront/scripts/preprocess_floorplan.py`, and make sure your custom floor plans dimensions are within x-y bounds of the training data. Other information stored in `dataset_stats.txt` is just for note keeping.

The `boxes.npz` file for each room can be compiled in a similar way as in the preprocessing script [here](https://github.com/MIT-SPARK/ThreedFront/blob/main/scripts/preprocess_data.py#L232-L247). This file contains three types of data: (1) object geometric features `class_labels, translations, sizes, angles`; (2) floor plan features `room_layout, floor_plan_vertices, floor_plan_faces, floor_plan_centroid`; (3) other scene info. If you are only interested in floor-conditioned scene synthesis, type (1) and (3) data will not matter. However, you still need to include these keys to avoid dataloader error. [This function](https://github.com/MIT-SPARK/ThreedFront/blob/main/threed_front/datasets/threed_front.py#L307) shows how `boxes.npz` is loaded to the dataloader. For example, we set object geometric features to `np.empty((N, E), dtype=np.float32)` where `N` is `sample_num_points` in the config file and `E` is the feature dimension, and we set everything else to None. 

For the floor plan features, there are two options. The first option is to format your custom floor plan as a mesh (i.e. provide `floor_plan_vertices, floor_plan_faces, floor_plan_centroid`), similar to 3D-FRONT, and use simple_3dviz to render it to a 2D binary mask (i.e. `room_layout`). Then, if you want to use PointNet, you can use [preprocess_floorplan.py](https://github.com/MIT-SPARK/ThreedFront/blob/main/scripts/preprocess_floorplan.py) to update your custom data directory (i.e. add `floor_plan_boundary_points_normals`). The other option is to compute `floor_plan_boundary_points_normals` directly and save to `boxes.npz`. You can start from ordered floor plan corners (see [this code](https://github.com/MIT-SPARK/ThreedFront/blob/main/scripts/preprocess_floorplan.py#L39-L41)). However, you cannot use ResNet feature extractor as it requires `room_layout`. Also, you cannot render results through simple_3dviz, since it only works with mesh.

Here are more details about the first option. You need to first format your custom floor plan as `floor_plan_vertices, floor_plan_faces, floor_plan_centroid` (i.e. as mesh). Below is an example code assuming `floor_plan_centroid=np.zeros(3)`:
```
from threed_front.datasets.threed_front_scene import rotation_matrix_around_y

def get_polygon(n):
    """ return a regular n-sided polygon with +y axis normal and centered at (0, 0, 0) """
    R_y = rotation_matrix_around_y(2*np.pi/n)
    vertices = np.zeros((3*n, 3))
    faces = np.arange(0, 3*n).reshape(n, 3)

    start_point = np.array([1, 0, 0])
    for j in range(n):
        next_point = R_y @ start_point
        vertices[j*3] = next_point
        vertices[j*3+2] = start_point
        start_point = next_point
    
    return vertices, faces
```
Then you use `threed_front.rendering.render_projection` to render the floor plan to a 2D mask. See example [here](https://github.com/MIT-SPARK/ThreedFront/blob/main/scripts/preprocess_data.py#L224-L228). Note that `get_floor_plan` takes in a `CachedRoom` object and it only uses `floor_plan` and `floor_plan_centroid` attributes. You can initialize `room = CachedRoom(None, None, floor_plan_vertices, floor_plan_faces, floor_plan_centroid, None, None, None, None, None, None, None)` and use it as input. Make sure "room_mask.png" is saved to the correct room_i subdirectory. These should be everything you need to provide in `boxes.npz`. Finally, you need to run preprocess_floorplan.py if you want to use PointNet.

**Comment by suny194** [comment]
> We have tested MiDiffusion with custom floor plan using PointNet feature extractor, but it should also work with ResNet.
> 
> The easiest way is to format the input data (which can contain just the floor plan and no objects) in the same way as preprocessed 3D-FRONT datasets. Then you can update `dataset_directory` in the config file to a `<custom_data_dir>` and run `scripts/generate_results.py <weight_file> --config_file config/<your_updated_config>.yaml`. Your custom data directory should look like this:
> 
> ```
> <ThreedFront_dir>/output/3d_front_processed/<custom_data_dir>/
> ├── dataset_stats.txt
> ├── room_1/
>     ├── boxes.npz
>     ├── room_mask.png
> ├── room_2/
>     ├── boxes.npz
>     ├── room_mask.png
> ```
> 
> You should copy `dataset_stats.txt` file from the preprocessed 3D-FRONT dataset directory (i.e. training data). This file contains upper and lower bounds for all geometric features (e.g. "bounds_translations"). These bounds are used to normalize raw input data and de-scale the predicted results. Hence, you need to pay attention to `"bounds_fpbpn"` ([x_min, y_min, dx_min, dy_min, x_max, y_max, dx_max, dx_max]), which is computed by `ThreedFront/scripts/preprocess_floorplan.py`, and make sure your custom floor plans dimensions are within x-y bounds of the training data. Other information stored in `dataset_stats.txt` is just for note keeping.
> 
> The `boxes.npz` file for each room can be compiled in a similar way as in the preprocessing script [here](https://github.com/MIT-SPARK/ThreedFront/blob/main/scripts/preprocess_data.py#L232-L247). This file contains three types of data: (1) object geometric features `class_labels, translations, sizes, angles`; (2) floor plan features `room_layout, floor_plan_vertices, floor_plan_faces, floor_plan_centroid`; (3) other scene info. If you are only interested in floor-conditioned scene synthesis, type (1) and (3) data will not matter. However, you still need to include these keys to avoid dataloader error. [This function](https://github.com/MIT-SPARK/ThreedFront/blob/main/threed_front/datasets/threed_front.py#L307) shows how `boxes.npz` is loaded to the dataloader. For example, we set object geometric features to `np.empty((N, E), dtype=np.float32)` where `N` is `sample_num_points` in the config file and `E` is the feature dimension, and we set everything else to None.
> 
> For the floor plan features, there are two options. The first option is to format your custom floor plan as a mesh (i.e. provide `floor_plan_vertices, floor_plan_faces, floor_plan_centroid`), similar to 3D-FRONT, and use simple_3dviz to render it to a 2D binary mask (i.e. `room_layout`). Then, if you want to use PointNet, you can use [preprocess_floorplan.py](https://github.com/MIT-SPARK/ThreedFront/blob/main/scripts/preprocess_floorplan.py) to update your custom data directory (i.e. add `floor_plan_boundary_points_normals`). The other option is to compute `floor_plan_boundary_points_normals` directly and save to `boxes.npz`. You can start from ordered floor plan corners (see [this code](https://github.com/MIT-SPARK/ThreedFront/blob/main/scripts/preprocess_floorplan.py#L39-L41)). However, you cannot use ResNet feature extractor as it requires `room_layout`. Also, you cannot render results through simple_3dviz, since it only works with mesh.
> 
> Here are more details about the first option. You need to first format your custom floor plan as `floor_plan_vertices, floor_plan_faces, floor_plan_centroid` (i.e. as mesh). Below is an example code assuming `floor_plan_centroid=np.zeros(3)`:
> 
> ```
> from threed_front.datasets.threed_front_scene import rotation_matrix_around_y
> 
> def get_polygon(n):
>     """ return a regular n-sided polygon with +y axis normal and centered at (0, 0, 0) """
>     R_y = rotation_matrix_around_y(2*np.pi/n)
>     vertices = np.zeros((3*n, 3))
>     faces = np.arange(0, 3*n).reshape(n, 3)
> 
>     start_point = np.array([1, 0, 0])
>     for j in range(n):
>         next_point = R_y @ start_point
>         vertices[j*3] = next_point
>         vertices[j*3+2] = start_point
>         start_point = next_point
>     
>     return vertices, faces
> ```
> 
> Then you use `threed_front.rendering.render_projection` to render the floor plan to a 2D mask. See example [here](https://github.com/MIT-SPARK/ThreedFront/blob/main/scripts/preprocess_data.py#L224-L228). Note that `get_floor_plan` takes in a `CachedRoom` object and it only uses `floor_plan` and `floor_plan_centroid` attributes. You can initialize `room = CachedRoom(None, None, floor_plan_vertices, floor_plan_faces, floor_plan_centroid, None, None, None, None, None, None, None)` and use it as input. Make sure "room_mask.png" is saved to the correct room_i subdirectory. These should be everything you need to provide in `boxes.npz`. Finally, you need to run preprocess_floorplan.py if you want to use PointNet.

Hi, I also tried this. The mask.png corresponding to npz has been generated. When updating config.yaml, in addition to dataset_directory
, should other parameters also be modified, such as dataset_type
? Or do more initialization operations need to be performed? Thank you for your time

**Comment by SiyiHu** [comment]
Thanks for the follow up questions! There are a few details missing in my first comment.

Yes, you will need to update "dataset_directory" and "annotation_file". This should be a CSV file similar to `bedroom_threed_front_splits.csv` [here](https://raw.githubusercontent.com/MIT-SPARK/ThreedFront/refs/heads/main/dataset_files/bedroom_threed_front_splits.csv). Since your custom data is only for testing, you can label everything as test data. The dataloader will load `boxes.npz` by finding data subdirectories included in the test split (see [here](https://github.com/MIT-SPARK/ThreedFront/blob/main/threed_front/datasets/threed_front.py#L268-L269)). For example, in my toy example where there are two custom floor plans saved to "room_1" and "room_2", the split file should be
```
1,train
1,test
2,test
```
The first line avoids IndexError when loading an "empty" `raw_train_dataset` [here](https://github.com/MIT-SPARK/MiDiffusion/blob/main/scripts/generate_results.py#L129-L134). 

By default, "dataset_directory" and "annotation_file" are in `<ThreedFront_dir>/output/3d_front_processed/` and `<ThreedFront_dir>/dataset_files/`, respectively (see `update_data_file_paths` which expands the full file paths [here](https://github.com/MIT-SPARK/MiDiffusion/blob/main/scripts/utils.py#L30-L35)).

If you want to use our pre-trained models, note that these models are trained with PointNet floor plan feature extractor. You will need to run `ThreedFront/scripts/preprocess_floorplan.py` to sample floor plan boundary. This script will over-write `bounds_fpbpn` in "datasets_stats.txt". Either disable that or copy "datasets_stats.txt" again.

You can then generate results using `scripts/generate_results.py`, and render with `scripts/render_results.py` (you might need to manually set "room_type" to model type to load the corresponding 3D-FUTURE CAD models).

**Comment by foojayyy** [comment]
Thank you so much for your help.I followed your instructions, and everything worked fine with my custom floor plans.
